clar_count.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" (C) Copyright 2019 Evils <evils.devils@protonmail.com>,
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH CLAR_COUNT 1 "2019-05-12"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
Clar_count \-
.B Clarissa
utility count known and unknown MAC addresses.
.SH SYNOPSIS
.B clar_count
.RI [path\ to\ macs.csv]\ [path\ to\ \fBClarissa\fR's\ list]\ [option]
.br
.SH DESCRIPTION
.B Clar_count
is a utility intended to count known and unknown MAC addresses from a macs.csv file, in \fBClarissa\fR's output list.
.br
It can output in a variety of formats specified by the option used.
.br
.SH OPTIONS
\fBClar_count\fR only takes 1 option per instance.
.br
These control it's output.
.br
Several can be parsed by \fBTelegraf\fR for use with \fBInfluxDB\fR.

.TP
.B -c, --csv
CSV with a header line (works with \fBTelegraf\fR)
.TP
.B -t, --title
One-line CSV containing the count of known addresses and of everything else.
.TP
.B -j, --json
JSON object (works with \fBTelegraf\fR)
.TP
.B -i, --influx
influx line protocol (works with \fBTelegraf\fR and presumably \fBInfluxDB\fR)
.TP
.B -n, --names
list of names, one name per line
.TP
.B -nj, --names_json
same content as -n but in a JSON array
.TP
.B -l, --log [salt]
Keep a hashed and salted log of unique addresses seen in the user's home directory.
.br
Unless a salt is supplied, a random salt will be used per instance.
.SH MAC.CSV FORMAT
MAC address,name
.br
MAC address,!name
.br
MAC address,name?
.br
MAC address,!name?
.br

Any lines without a MAC address won't get used and any fields after the name will be ignored.
.br
Any name with an exclamation mark
.B !
will not be output,
.br
and any name containing a question mark
.B ?
will not be counted.
.SH SEE ALSO
.BR clarissa(8),
.BR telegraf,
.BR influxdb
.br

https://gitlab.com/evils/clarissa


debian/changelog
clarissa (0.6) unstable; urgency=medium

  * use the Makefile for debian
  * clean up Makefile to work with nix
  * read-only support for the any device
  * prevent the loopback device from being selected
  * add SIGTERM support

 -- Evils <evils.devils@protonmail.com>  Thu, 26 Sep 2019 23:40:32 +0200

clarissa (0.5) unstable; urgency=medium

  * change default output filename to describe the used options
  * switch to pcap_findalldevs() (pcap_lookupdev() obsoletion)

 -- Evils <evils.devils@protonmail.com>  Fri, 19 Jul 2019 22:45:50 +0200

clarissa (0.4) unstable; urgency=medium

  * fix dump_state() memory leak
  * catch SIGINT, exit loop and clean up
  * "any" device listening support (no sending)
  * fix procd init script to set interface instead of interval...

 -- Evils <evils.devils@protonmail.com>  Fri, 31 May 2019 08:11:40 +0200

clarissa (0.3) unstable; urgency=medium

  * add long options
  * add README.md
  * fix pcap_next() hanging

 -- Evils <evils.devils@protonmail.com>  Thu, 16 May 2019 07:51:55 +0200

clarissa (0.2.1) unstable; urgency=medium

  * bug patch, minimum timeout got hampered by pcap_next() hanging,
    adding a few ms delay to the main loop inexplicably mitigates this.

  * adjusted defaults

 -- Evils <evils.devils@protonmail.com>  Wed, 15 May 2019 04:59:25 +0200

clarissa (0.2) unstable; urgency=medium

  * working NDP Neighbor Solicitation (IPv6 nagging)

 -- Evils <evils.devils@protonmail.com>  Sun, 12 May 2019 02:16:31 +0200

clarissa (0.1) unstable; urgency=medium

  * Initial packaging

 -- Evils <evils.devils@protonmail.com>  Tue, 30 Apr 2019 02:10:14 +0200


clar_count.sh
#!/usr/bin/env bash

# get the count of MAC addresses named in the macs.csv ($1) for all the found MAC addresses ($2)
# and output a count of the unique names (minus those with "?") and tally of the remaining addresses
# and provide a list of those names (minus those with "!")
# in a variety of formats (specified by $3)

# the format options are:
# -c or --csv, CSV with header (accepted by telegraf (for influxdb))
# -t or --title, short CSV showing the count of known entities and everything else
# -j or --json, json format (accepted by telegraf (for influxdb))
# -i or --influx, influx line protocol (accepted by telegraf (for influxdb))
# -n or --names, show all names minus those with an exclamation mark "!"
# -nj or --names_json, same as --names but in a json array


# maybe set this to measurement's location? (useful for influxdb)
NAME="clarissa"

# show correct usage if used incorrectly
if [ -z "$3" ]; then
echo "Please use the following format:"
echo "clar_count.sh [path to macs.csv] [path to clarissa's output] [option (-a for all)]"
fi

# do the actual counting

NAMES=()
TALLY=0

while read -r; do
	LINE="$(grep -i "$REPLY" "$1")"
	if [ "$LINE" ]; then
		NAMES+=("$(echo "$LINE" | awk -F "," '{print $2}')" )
	else
		(( TALLY++ ))
	fi
done < "$2"

COUNT="$(printf '%s\n' "${NAMES[@]}" | sed -e '/^\s*$/d' -e '/[?‽]/d' | sort | uniq | wc -l)"


# formatted counts as:

csv() {
	echo "name,counted,balance"
	echo "$NAME"",""$COUNT"",""$TALLY"
}

title() {
	echo "$COUNT"",""$TALLY"
}

json() {
echo "{\"name\":\"""$NAME""\", \"counted\":""$COUNT"", \"balance\":""$TALLY""}"
}

influx() {
echo "$NAME"" counted=""$COUNT"",balance=""$TALLY"
}

names_json() {
names | jq -csR '[ split ("\n") | .[] | select(length > 0)]'
}

names() {
printf '%s\n' "${NAMES[@]}" | sed -e 's/?//g' -e '/[!‽]/d' | sort | uniq
}

# log the number of unique devices seen during a running instance of this function
log() {

salt_n=420
if [ ! "$2" ]; then
	salt="$(head -c "$salt_n" /dev/urandom | hexdump -v)"
	echo "Using random salt of length: $salt_n"
else
	salt="$(echo "$2" | hexdump -v)"
	echo "Using salt: $salt"
fi

log=~/clar_"$(date +%s)".log

echo
echo "Session unique hashes stored in $log"

while true; do

	while read -r; do

		echo "$REPLY" "$salt" | sha256sum >> "$log"

	done < "$1"

	sort "$log" | uniq > "$log.tmp"
	mv "$log.tmp" "$log"

	sleep 60

done

}


# handle the options

case "$3" in

	-c|--csv) csv ;;

	-t|--title) title ;;

	-j|--json) json ;;

	-i|--influx) influx ;;

	-n|--names) names ;;

	-nj|-jn|--names_json) names_json ;;

	-l|--log) log "$2" "$4" ;;

	# for testing
	-a|--all) csv; echo; title; echo;  json; echo; influx; echo; names_json; echo; names ;;

	*) exit 1 ;;
esac


# the format of macs.csv should be:
# MAC address,name
# MAC address,!name
# MAC address,name?
# MAC address,!name?

# any lines without a MAC address won't get used and any fields after the name will be ignored
# additionally, any name with an exclamation mark ("!") will not be shown by names()
# and any name containing a question mark ("?") will not be counted in $COUNT

# NOTE, while having an empty field for the display name column currently hase the same effect as having both "!" and "?", this could be subject to change
